### PR TITLE
Cling visibility and semantic interposition

### PR DESCRIPTION
**Pull request by Axel-Naumann**
Wit kudos to @hahnjo !

### CODE REVIEW

**Comment by vgvassilev** [comment]
Yes! Could you also now try running clingtest as part of the regular ROOT testing procedure? I believe (with minor adjustments) it will work.

File: interpreter/CMakeLists.txt [hunk 1]
```diff
@@ -131,6 +131,10 @@ else()
   set(CMAKE_CXX_VISIBILITY_PRESET hidden)
   set(CMAKE_C_VISIBILITY_PRESET hidden)
 endif()
+if (CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
+  ROOT_ADD_CXX_FLAG(CMAKE_CXX_FLAGS "-fno-semantic-interposition")
+endif()
```
Comment (hahnjo) [inline interpreter/CMakeLists.txt:136]: We likely also want this for Clang, right? (or is `CMAKE_CXX_COMPILER_ID` set to `GNU` in that case as well?)

Comment (Axel-Naumann) [reply interpreter/CMakeLists.txt]: I found this restriction in https://reviews.llvm.org/D102453:

> Note: we cannot add -fno-semantic-interposition for Clang. Clang's implementation additionally optimizes global variables, which is incompatible with unfortunate ELF -fno-pic default: direct access relocations for external data. If the executable has a -fno-pic object file referencing a global variable declared in a public header, the direct access relocation will cause a copy relocation. The executable and libLLVM.so/libclang-cpp.so will disagree on the address.

OTOH why not try :-)

Comment (hahnjo) [reply interpreter/CMakeLists.txt]: Ugh, unfortunate. On the other hand, we don't want users to call into LLVM directly so we should be fine?

**Comment by Axel-Naumann** [comment]
@phsft-bot build with flags -Dclingtest=On

**Review by hahnjo — APPROVED** [review]
:+1: (maybe we should consider removing `countIncludePaths`, I also see this failing locally in certain configurations...)

**Comment by Axel-Naumann** [comment]
@hahnjo `countIncludePaths` is fine to fail with `-Dclingtest=On` - that's exactly what we're trying to catch with this test :-) We could filter a bit (better), maybe? In the end what we want to catch with this test is CMake changes (e.g. during upgrade) or rootcling changes that inject a load of irrelevant / wrong `-I`s. We had an instance of that in the past.

**Comment by Axel-Naumann** [comment]
@phsft-bot build

**Comment by Axel-Naumann** [comment]
@phsft-bot build with flags -Dclingtest=On

**Comment by Axel-Naumann** [comment]
Macs were failing exactly like this before (in numba, with `-Dclingtest=On`): https://github.com/root-project/root/pull/3760#issuecomment-847791104

**Review by vgvassilev — APPROVED** [review]
LGTM!

**Review by oshadura — APPROVED** [review]
CMake changes makes sense to me, thanks for improvements!

**Pull request merged**